pkg/resource/stack/deployment.go

```diff
@@ -240,43 +252,16 @@ func DeserializeDeploymentV3(
 	}
 
 	var dec config.Decrypter
-	if secretsManager == nil {
-		var ciphertexts []string
-		for _, res := range deployment.Resources {
-			collectCiphertexts(&ciphertexts, res.Inputs)
-			collectCiphertexts(&ciphertexts, res.Outputs)
-		}
-		if len(ciphertexts) > 0 {
-			// If there are ciphertexts, but we couldn't set up a secrets manager, error out early
-			// to avoid panic'ing later on.  This snapshot is broken and needs to be repaired
-			// manually.
-			return nil, errors.New("snapshot contains encrypted secrets but no secrets manager could be found")
-		}
-		dec = config.NewPanicCrypter()
+	if secretsManager != nil {
+		dec = secretsManager.Decrypter()
 	} else {
-		d := secretsManager.Decrypter()
-
-		// Do a first pass through state and collect all of the secrets that need decrypting.
-		// We will collect all secrets and decrypt them all at once, rather than just-in-time.
-		// We do this to avoid serial calls to the decryption endpoint which can result in long
-		// wait times in stacks with a large number of secrets.
-		var ciphertexts []string
-		for _, res := range deployment.Resources {
-			collectCiphertexts(&ciphertexts, res.Inputs)
-			collectCiphertexts(&ciphertexts, res.Outputs)
-		}
+		dec = config.NewErrorCrypter("snapshot contains encrypted secrets but no secrets manager could be found")
+	}
 
-		// Decrypt the collected secrets and create a decrypter that will use the result as a cache.
-		decrypted, err := d.BatchDecrypt(ctx, ciphertexts)
-		if err != nil {
-			return nil, err
-		}
-		contract.Assertf(len(decrypted) == len(ciphertexts), "decrypted secrets count does not match ciphertexts count")
-		cache := make(map[string]string)
-		for i, ciphertext := range ciphertexts {
-			cache[ciphertext] = decrypted[i]
-		}
-		dec = newMapDecrypter(d, cache)
+	var completeBatch CompleteCrypterBatch
+	if batchingSecretsManager, ok := secretsManager.(BatchingSecretsManager); ok {
```

Comment:
Should this be inside the `if` above? It mirrors the encryption flow then, right?

Reply:
I don't think it's functionally different, but yes, putting it inside would be easier to visually mirror the serialize method flow 👍